README.md
<img align = 'center' src="https://github.com/Butcanudothis/thebettertodoflutter/blob/master/Screenshot%202023-05-05%20at%207.41.02%20am.png" height ="720" width ="360">

# The Better Todo Flutter

The Better Todo Flutter is a todo list app that aims to provide a smooth and seamless user experience. The app is built using Flutter and features state management with the Provider package and Firebase Auth for Google Sign-In.

### Features
- Add, delete, and check off tasks
- Google Sign-In using Firebase Auth
- Support for both light and dark themes
- Smooth and seamless user experience
- Simple and easy-to-use interface

### Getting Started
To get started with The Better Todo Flutter, follow these steps:

1. Clone the repository onto your local machine:
```bash
git clone https://github.com/username/the-better-todo-flutter.git
```

2. Install the necessary packages using `flutter pub get`

3. Run the app on your device or emulator using `flutter run`

### Changelog
For a complete list of changes and updates, please see the [changelog](CHANGELOG.md).

### Contributing
Contributions to The Better Todo Flutter are welcome and encouraged. To contribute, please follow these steps:

1. Fork the repository
2. Create a new branch for your feature or bugfix
3. Commit your changes to your new branch
4. Create a pull request


CHANGELOG.md
## Changelog

### 1.0.0 (Oct 5, 2020)
- Added Google Sign-In functionality using Firebase Auth
- Added un-versioned files
- Updated README.md

### 0.9.0 (Sep 28, 2020)
- Added Deletion functionality
- Final refactoring using Provider Package State Management

### 0.8.0 (Sep 24, 2020)
- Completed basic functionality

### 0.7.0 (Sep 22, 2020)
- Removed TaskCheckBox and other stateful widgets

### 0.6.0 (Sep 20, 2020)
- Theme overhaul

### 0.5.0 (Sep 19, 2020)
- Added callbacks and checkBox state
- Added task check functionality using local state

### 0.4.0 (Sep 18, 2020)
- Added task screen

### 0.3.0 (Sep 16, 2020)
- Added onPressed

### 0.2.0 (Sep 14, 2020)
- Refactored ListView and ListTile
- Basic layout

### 0.1.0 (Sep 2, 2020)
- Initial commit


lib/widgets/task_list.dart
import 'package:flutter/material.dart';
import 'task_tile.dart';
import 'package:flutter_todo/models/task_data.dart';
import 'package:provider/provider.dart';

class TaskList extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Consumer<TaskData>(
      builder: (context, taskData, child) {
        return ListView.builder(
          padding: EdgeInsets.fromLTRB(20, 25, 20, 10),
          itemBuilder: (context, index) {
            final task = taskData.tasks[index];
            return TaskTile(
                isChecked: task.isDone,
                taskTitle: task.name,
                checkboxCallback: (checkboxState) {
                  taskData.updateTask(task);
                },
                deletionCallback: () {
                  taskData.deleteTask(task);
                });
          },
          itemCount: taskData.countTasks,
        );
      },
    );
  }
}


lib/widgets/task_tile.dart
import 'package:flutter/material.dart';

class TaskTile extends StatelessWidget {
  final bool isChecked;
  final String taskTitle;
  final Function checkboxCallback;
  final Function deletionCallback;
  TaskTile(
      {this.isChecked,
      this.taskTitle,
      this.checkboxCallback,
      this.deletionCallback});
  @override
  Widget build(BuildContext context) {
    return ListTile(
      onLongPress: deletionCallback,
      title: Text(
        taskTitle,
        style: TextStyle(
            color: Colors.orangeAccent,
            fontWeight: FontWeight.w500,
            decoration: isChecked ? TextDecoration.lineThrough : null),
      ),
      trailing: Checkbox(
        activeColor: Colors.orangeAccent,
        checkColor: Colors.blueGrey.shade900,
        value: isChecked,
        onChanged: checkboxCallback,
      ),
    );
  }
}

// class TaskCheckbox extends StatelessWidget {
//   final bool checkboxState;
//   final Function toggleCheckboxState;
//   TaskCheckbox({this.checkboxState, this.toggleCheckboxState});
//   @override
//   Widget build(BuildContext context) {
//     return
//   }
// }


lib/screens/taskscreen.dart
import 'package:flutter/material.dart';
import 'package:flutter/painting.dart';
import 'package:flutter_todo/screens/add_task_screen.dart';
import 'package:flutter_todo/widgets/task_list.dart';
import 'package:flutter_todo/models/task_data.dart';
import 'package:provider/provider.dart';
import 'package:firebase_auth/firebase_auth.dart';

class TasksScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.orangeAccent,
      floatingActionButton: FloatingActionButton(
        backgroundColor: Colors.orangeAccent,
        child: Icon(
          Icons.add,
          color: Colors.blueGrey.shade900,
        ),
        elevation: 1.5,
        onPressed: () {
          showModalBottomSheet(
              context: context,
              builder: (BuildContext context) => AddTaskScreen(
                  // addTaskCallback: (newTaskTitle) {
                  //   setState(() {
                  //     //  tasks.add(Task(name: newTaskTitle));
                  //   });
                  //   Navigator.pop(context);
                  // },
                  ));
        },
      ),
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Container(
            padding:
                EdgeInsets.only(left: 30, right: 30.0, top: 60, bottom: 30.0),
            child: Container(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: <Widget>[
                  FlatButton(
                    onPressed: () {
                      FirebaseAuth.instance.signOut();
                      Navigator.pop(context);
                    },
                    child: Hero(
                      tag: "icon",
                      child: CircleAvatar(
                        backgroundColor: Colors.blueGrey.shade900,
                        radius: 30.0,
                        child: Icon(
                          Icons.track_changes,
                          color: Colors.orangeAccent,
                          size: 40,
                        ),
                      ),
                    ),
                  ),
                  SizedBox(
                    height: 10.0,
                  ),
                  Text(
                    'Just do it.',
                    style: TextStyle(
                      color: Colors.blueGrey.shade900,
                      fontSize: 40,
                      fontWeight: FontWeight.w700,
                    ),
                  ),
                  Text(
                    '${Provider.of<TaskData>(context).countTasks} Tasks',
                    style: TextStyle(
                        color: Colors.blueGrey.shade900,
                        fontSize: 18.0,
                        fontWeight: FontWeight.w700),
                  ),
                ],
              ),
            ),
          ),
          Expanded(
            child: Container(
              height: 300,
              decoration: BoxDecoration(
                color: Colors.blueGrey.shade900,
                borderRadius: BorderRadius.only(
                  topRight: Radius.circular(20.0),
                  topLeft: Radius.circular(20.0),
                ),
              ),
              child: TaskList(),
            ),
          ),
        ],
      ),
    );
  }
}


lib/screens/add_task_screen.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_todo/models/task_data.dart';
import 'package:provider/provider.dart';

class AddTaskScreen extends StatelessWidget {
  //final Function addTaskCallback;

  //AddTaskScreen({this.addTaskCallback});
  @override
  Widget build(BuildContext context) {
    String newTaskTitle;
    return Container(
      color: Color(0xFF101010),
      child: Container(
        padding: EdgeInsets.fromLTRB(35, 10, 35, 60),
        decoration: BoxDecoration(
          color: Colors.orangeAccent,
          borderRadius: BorderRadius.only(
            topLeft: Radius.circular(20.0),
            topRight: Radius.circular(20.0),
          ),
        ),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Text(
              'Add Task',
              textAlign: TextAlign.center,
              style: TextStyle(
                  color: Colors.blueGrey.shade900,
                  fontSize: 30.0,
                  fontWeight: FontWeight.w700),
            ),
            TextField(
              onChanged: (newText) {
                newTaskTitle = newText;
              },
              textAlign: TextAlign.center,
              autofocus: true,
              style: TextStyle(
                fontSize: 20,
              ),
            ),
            FlatButton(
              onPressed: () {
                Provider.of<TaskData>(context, listen: false)
                    .addTask(newTaskTitle);
                //addTaskCallback(newTaskTitle);
                Navigator.pop(context);
              },
              color: Colors.blueGrey.shade900,
              child: Text(
                'Add',
                style: TextStyle(color: Colors.orangeAccent, fontSize: 25.0),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/screens/login_screen.dart
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:google_sign_in/google_sign_in.dart';
import 'taskscreen.dart';
import 'package:firebase_core/firebase_core.dart';

Future<UserCredential> signInWithGoogle() async {
  // Trigger the authentication flow
  final GoogleSignInAccount googleUser = await GoogleSignIn().signIn();

  // Obtain the auth details from the request
  final GoogleSignInAuthentication googleAuth = await googleUser.authentication;

  // Create a new credential
  final GoogleAuthCredential credential = GoogleAuthProvider.credential(
    accessToken: googleAuth.accessToken,
    idToken: googleAuth.idToken,
  );

  // Once signed in, return the UserCredential
  return await FirebaseAuth.instance.signInWithCredential(credential);
}

class LoginScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.orangeAccent,
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.center,
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          Hero(
            tag: "icon",
            child: CircleAvatar(
              backgroundColor: Colors.blueGrey.shade900,
              radius: 60.0,
              child: Icon(
                Icons.track_changes,
                color: Colors.orangeAccent,
                size: 70,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Container(
              height: 300,
              width: 400,
              padding: EdgeInsets.symmetric(horizontal: 20, vertical: 30),
              decoration: BoxDecoration(
                color: Colors.blueGrey.shade900,
                borderRadius: BorderRadius.all(
                  Radius.circular(20.0),
                ),
              ),
              child: Column(
                children: [
                  FlatButton(
                    onPressed: () {
                      //GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                      //                               .requestEmail()
                      //                               .build();
                      signInWithGoogle();
                      Firebase.initializeApp();
                      Navigator.push(
                          context,
                          MaterialPageRoute(
                              builder: (BuildContext context) =>
                                  TasksScreen()));
                    },
                    child: Container(
                      decoration: BoxDecoration(
                          color: Colors.transparent,
                          borderRadius: BorderRadius.circular(30),
                          border: Border.all(
                            color: Colors.orangeAccent,
                            width: 3,
                          )),
                      child: Padding(
                        padding: const EdgeInsets.symmetric(
                            vertical: 8.0, horizontal: 100.0),
                        child: Text(
                          'Login',
                          style: TextStyle(
                              fontWeight: FontWeight.w400,
                              fontSize: 30,
                              color: Colors.orangeAccent),
                        ),
                      ),
                    ),
                  )
                ],
              ),
            ),
          )
        ],
      ),
    );
  }
}
